sdks/python/apache_beam/runners/worker/statesampler.pyx

```diff
@@ -150,11 +161,31 @@ cdef class StateSampler(object):
           # without the GIL.
           nsecs_ptr = &(<ScopedState>PyList_GET_ITEM(
               self.scoped_states_by_index, self.current_state_index)).nsecs
+          lull_nsecs_ptr = &(<ScopedState>PyList_GET_ITEM(
+              self.scoped_states_by_index, self.current_state_index)).lull_nsecs
+
           nsecs_ptr[0] += elapsed_nsecs
+          lull_nsecs_ptr[0] += elapsed_nsecs
+
+          if lull_nsecs_ptr[0] > self.lull_threshold_ms * MSECS_TO_NSECS:
+            report_lull = True
+            lull_state = self.current_state_index
+
           last_nsecs += elapsed_nsecs
         finally:
           pythread.PyThread_release_lock(self.lock)
 
+        if report_lull:
+          with gil:
```

Comment:
<!--thread_id:cc_143091728_t; commit:95830a027670626ce0f081991d56dac60c8ee8d1; resolved:0-->
<!--section:context-quote-->
> **pabloem** wrote:
> To avoid the deadlock, we wait to release the sampler lock in line 176, and acquire the GIL in line 179. That guarantees no deadlocks.

<!--section:body-->
Right, that would be one way to do it--however, as Robert pointed out, this would result in a performance hit, since this code path is executed on each transition.  We shouldn't check in code that we know could result in a deadlock.

I was going to suggest the path of (1) logging to stderr here and (2) doing something like pushing to a GIL-less list and calling logging.* in another thread, but one other path you could consider is populating information how much time we have spent in the current state (in the `StateSamplerInfo` object returned by `StateSampler.get_info()`) and reporting this in the progress reporting thread if the transition count hasn't changed past a threshold (search for `progress_reporting_thread` in `dataflow_worker`).  This would result in minimal change to the code here and avoid these messy locking issues.